cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java

```diff
@@ -75,20 +75,36 @@ public void run() {
       try {
         DatasetId v2MetricsTableId = getMetricsDatasetId(v2TableNamePrefix, resolution);
         DatasetId v3MetricsDatasetId = getMetricsDatasetId(v3TableNamePrefix, resolution);
-        if (MigrationTableHelper.hasInstanceWithRetry(datasetFramework, v2MetricsTableId)) {
+        Boolean hasInstance =  MigrationTableHelper.hasInstanceWithRetry(datasetFramework, v2MetricsTableId);
+
+        if (hasInstance == null) {
+          // hasInstance returns null during stop; so safe to break
+          break;
+        }
+
+        if (hasInstance) {
           try (MetricsTable v2MetricsTable =
                  MigrationTableHelper.getDatasetWithRetry(datasetFramework, v2MetricsTableId);
                MetricsTable v3MetricsTable =
                  MigrationTableHelper.getDatasetWithRetry(datasetFramework, v3MetricsDatasetId)) {
+            if (v2MetricsTable == null || v3MetricsTable == null) {
+              // we get null when we are stopping, so safe to break
+              break;
+            }
             MetricsTableMigration metricsTableMigration = new MetricsTableMigration(v2MetricsTable, v3MetricsTable);
             metricsTableMigration.transferData(sleepMillisBetweenTransfer);
             LOG.info("Metrics table data migration is complete for resolution {}", resolution);
             // don't delete if we are stopping
-            if (!stopping) {
-              // now transfer is complete; its safe to delete the v2 metrics table
-              MigrationTableHelper.deleteInstanceWithRetry(datasetFramework, v2MetricsTableId);
+            if (stopping) {
+              break;
+            }
+            // now transfer is complete; its safe to delete the v2 metrics table
+            if (MigrationTableHelper.deleteInstanceWithRetry(datasetFramework, v2MetricsTableId)) {
```

Comment:
it may be better to throw InterruptedException from the retryable methods to indicate that your're stopping. You are already handling that exception correctly.